### PR TITLE
Initial code and CI for trace data pipeline

### PR DESCRIPTION
**Pull request by bantonsson**
# What does this PR do?

The work done on a shared component data pipeline done during the R&D week.

# Motivation

Add a base build and CI that we can continue to iterate on to refine the API and functionality of the shared data pipeline component.

# Additional Notes

Please don't be to hard on the public APIs, and the implementation. They will evolve over time, but in the interest of not having a ginormous PR from a feature branch, we would like to get the machinery into main.

# How to test the change?

Describe here in detail how the change can be validated.

## For Reviewers
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Comment by paullegranddc** [comment]
Are we going to ship different shared libraries for data-pipeline and profiling?

**Comment by bantonsson** [comment]
We are not there yet, and in some languages we definitely will since they don't use the profiler in libdatadog.

**Comment by ivoanjo** [comment]
Not a blocker at this point, but more of an observation, note that shipping profiling AND data-pipeline separately for the same language will probably increase the binary size a lot more than if a unified libdatadog was produced ;)

**Comment by bantonsson** [comment]
Yes we are aware of the added binary size and are going to work for reducing this and combining the features into a single binary where it makes sense.

**Comment by paullegranddc** [comment]
> Not a blocker at this point, but more of an observation, note that shipping profiling AND data-pipeline separately for the same language will probably increase the binary size a lot more than if a unified libdatadog was produced ;)

It's also going to be hard to load both of them in the same process because of symbols exposed in ddcomon being in both shared library

**Comment by ivoanjo** [comment]
> It's also going to be hard to load both of them in the same process because of symbols exposed in ddcomon being in both shared library

Good point! [For Ruby we explicitly load the profiling native extension with `RTLD_LOCAL`](https://github.com/DataDog/dd-trace-rb/blob/master/ext/datadog_profiling_loader/datadog_profiling_loader.c#L68) and I believe that's a potential solution to this issue.

I'm saying this more out of curiosity / sharing notes :)

File: build-data-pipeline-ffi.sh [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env bash
+
+# Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+# SPDX-License-Identifier: Apache-2.0
+
+set -eu
+
+destdir="$1"
+
```
Comment (r1viollet) [inline build-data-pipeline-ffi.sh:9]: do we want to fix the build logics before adding the tracing logics or after ?

Comment (bantonsson) [reply build-data-pipeline-ffi.sh:9]: Could you please elaborate a bit on what this would mean.

Comment (r1viollet) [reply build-data-pipeline-ffi.sh:9]: This is relative to the RFC + issue https://github.com/DataDog/libdatadog/issues/353
The idea would be to have a single build where we can select what is required to build.
A good first step could be to unify the build scripts.
Ideally all logics would be embedded in cargo in the long term.
cargo build --feature trace-pipeline,profiling,telemetry

**Review by gleocadie — COMMENTED** [review]
LGTM
can you remove this change please windows/libdatadog.csproj
It will generate a +300MB nuget package. We need to rethink how we package this on windows. CI (dd-trace-dotnet for example, this can be a problem on the runners - not enough disk space)

File: build-data-pipeline-ffi.sh (hunk 1 above)
Comment (hoolioh) [reply build-data-pipeline-ffi.sh:9]: Fixed in the latest commit. As discussed internally profiling library will be the main artifact and data-pipeline symbols will be exported there.

File: profiling-ffi/src/lib.rs [hunk 2]
```diff
@@ -18,6 +18,10 @@ pub use crashtracker::*;
 #[cfg(feature = "ddtelemetry-ffi")]
 pub use ddtelemetry_ffi::*;
 
+#[cfg(feature = "data-pipeline-ffi")]
+#[allow(unused_imports)]
+pub use data_pipeline_ffi::*;
```
Comment (r1viollet) [inline profiling-ffi/src/lib.rs:23]: :+1:

File: build-data-pipeline-ffi.sh (hunk 1 above)
Comment (r1viollet) [reply build-data-pipeline-ffi.sh:9]: Why do we export the symbols here ?

Comment (hoolioh) [reply build-data-pipeline-ffi.sh]: Forgot the last commit (pushed it but the prompt asked for the credentials while I was AFK) now it's complete.

File: examples/ffi/trace_exporter.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+int main(int argc, char** argv)
+{
+    ddog_CharSlice host = DDOG_CHARSLICE_C("localhost");
+    uint16_t port = 8126;
+    ddog_CharSlice tracer_version = DDOG_CHARSLICE_C("v0.1");
+    ddog_CharSlice language = DDOG_CHARSLICE_C("dotnet");
+    ddog_CharSlice language_version = DDOG_CHARSLICE_C("10.0");
+    ddog_CharSlice language_interpreter = DDOG_CHARSLICE_C("X");
+    ddog_TraceExporter* trace_exporter = ddog_trace_exporter_new(
+        host,
```
Comment (r1viollet) [inline examples/ffi/trace_exporter.c:18]: we have an endpoint object that could be useful to share

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -97,7 +97,7 @@ jobs:
       - name: "Generate profiling FFI"
```
[the remark this answers is not in the document]
Comment (r1viollet) [reply .github/workflows/test.yml:97]: The important part is having at least one build per delivered combination.

**Review by r1viollet — APPROVED** [review]
LGTM
Minor suggestions. Exciting to see this progress :clap:

**Pull request merged**